File: W3/homeWork_W3/С++/Pen&Paper/Pen.cpp
#include <string>
#include "Pen.h"

using namespace std;

Pen::Pen(int inkCapacity) {
    this->inkAmount = 0;
    this->inkCapacity = inkCapacity;
}

Pen::~Pen() {
    
}

int Pen::getInkAmount() const {
    return this->inkAmount;
}

int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const string& message) {
    int start = 0;
    
    if ( this->inkAmount == 0 ) {
        throw OutOfInk();
    }
    string newMessage = message.substr(start, this->inkAmount);
    paper.addContent(newMessage);
    this->inkAmount = this->inkCapacity - paper.getSymbols();
}

void Pen::refill() {
    if ( this->inkCapacity <= 0 ) {
        cout << "You entered incorect pen capacity" << endl;
        return;  
    }
    this->inkAmount = this->inkCapacity;
}

File: W3/homeWork_W3/Java/PenPaper/Pen.java
public class Pen {
    private int inkAmount;
    private int inkCapacity;

    public Pen(int inkCapacity) {
        this.inkAmount = inkCapacity;
        this.inkCapacity = inkCapacity;
    }

    public Pen() {
        this(4096);
    }

    public final int getInkAmount() {
        return this.inkAmount;
    }
    public final int getInkCapacity() {
        return this.inkCapacity;
    }
    
    public void write(Paper paper, String message) throws OutOfInkException, OutOfSpaceException {
        if ( this.inkAmount == 0 ) {
            throw new OutOfInkException();
        }
        if ( message.length() > this.inkAmount ) {
            paper.addContent(message.substring(0, this.inkAmount));
            this.inkAmount = 0;
            return;
        }
        paper.addContent(message);
        this.inkAmount -= message.length();
    }
    
    public void refill() {
        if ( this.inkCapacity <= 0 ) {
            System.out.println("You entered incorect pen capacity");
            return;  
        }
        this.inkAmount = this.inkCapacity;
    }
}
